### ISSUE TITLE
[RFC] PT2-Friendly Traceable, Functional Collective Communication APIs

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#93173]
### 🚀 Traceable Collectives!

Collective APIs (e.g. all_reduce, all_gather, ...) are used in distributed PyTorch programs, but do not compose cleanly with compilers.  

Specifically, torchDynamo and the AotAutograd pipeline for decompositions and functionalization do not work with the existing c10d collective APIs
* there are not functional variants of these collectives
* ProcessGroup and Work objects interfere with graph tracing and pollute the IR with non-tensor objects

XLA also currently has to implement some workarounds, to marry the XLA collective ops via lazy tensor tracing with the existing PyTorch / C10D side. They have to use a custom ProcessGroup implementation and swizzle PTD PG creation functions. 

**Goals**
1) provide collectives that are **traceable** with the PT2 stack and XLA stack
2) provide **functional** collectives, which are easier for IR transformations to reason about
3) support **eager and compiled** flows with the same API
4) use **plain data types** in the traced API
5) allow tracing/compilation **without requiring process group init**
6) **support different frontends** (DTensors, ProcessGroups, etc)
7) support **autograd** for collective ops
8) clean up c10d python bindings and dispatcher registrations

**Non-goals**

1) Introduce multiple stream semantics in inductor

<img width="1530" alt="image" src="https://user-images.githubusercontent.com/4984825/216402330-b78e08ff-7d51-407b-aaee-f06aca9b60ec.png">

### New traceable collectives python API
```
def collective(input:Tensor, *, group: GROUP_TYPE) -> AsyncTensor
```
`GROUP_TYPE` is a Union over List<rank>, DeviceMesh, ProcessGroup, etc.  It allows flexible usage by different frontends.

`AsyncTensor` is a Tensor subclass that calls `wait()` automatically when the tensor is used by another op.

### New Dispatcher Collectives

```
aten::collective(Tensor, *, str tag, int[] ranks, int stride) -> Tensor`
```
These are the ops that actually get traced into a graph and can be manipulated by compiler passes.

The collective ops are functional, but compilers may be able to convert them to inplace.  They are asynchronous.

These ops support meta device (for traceability), and support backwards via derivatives.yaml.

The semantics of these ops are that they return a real tensor, but you aren't allowed to access its data or storage.

```
c10d.wait(Tensor) -> Tensor
```

`wait()` must be called on the output of any collective before its underlying data or storage is accessed.
* It is valid to peek at the size() or stride() (or probably other metadata) of a tensor returned from a collective, but not its data.  
* wait() is the only way to make an output from collectives safe to use by other non collective ops
* we are considering whether wait(collective(collective)) can be implemented safely, but by default we assume it is not
The semantics of wait are that you must only access the storage of the tensor returned from wait.  You can't think of wait as mutating its input tensor and making it safe to use.

### Alternatives

The following style of API has also been considered.  Its main disadvantage is in requiring a user to first initialize a processgroup, but it is also opaque and not easily interchangeable with lists of ranks or DTensors. It doesn't allow us to easily represent MPMD collectives.

```
pg = init_process_group()
pg_id = dist.register_process_group(pg)
collective(tensor, pg_id)
```

### Detailed Proposal
See [Traceable Collectives Design](https://docs.google.com/document/d/1Jqa68gvuVeFWZJFOiukmb58jAaUEET1GVMkd1GOMRT4/edit)

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @chauhang @penguinwu @zou3519 @bdhirsh @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @kwen2501 @XilunWu @ezyang @msaroufim @anijain2305 @soumith @ngimel

**Comment by alanwaketan** [comment]
Thanks, @wconstab. This is awesome! (Leaving a comment to have myself cc on the issue.)

**Comment by wconstab** [comment]
so one thing @ezyang pointed out is, we might not be able to support `list[list[]]` in our current codegen pipeline.  But if we can assume rectangular shape, we can just do `list[]` and have an extra `stride` int that lets us make the second dimension.  

But apart from that- is this api going to help for XLA? And is there anything wrong with it?

**Comment by ezyang** [comment]
Remind me what tag is?

**Comment by JackCaoG** [comment]
@wconstab Can you expand a bit on `do list[] and have an extra stride int that lets us make the second dimension`. Is it something like
[0,1,2,3,4,5,6,7] and for stride 2, we have 2 group [0,2,4,6] and [1,3,5,7]? It is a bit confusing TBH lol.

**Comment by ezyang** [comment]
Well, idk what Will has in mind, but I would instead have repped it the other way: [1,2,3,4,5,6,7,8], stride=4 gives me [[1,2,3,4],[5,6,7,8]]

**Comment by wconstab** [comment]
> Remind me what tag is?

@ezyang I'm not totally sure we need tag,  but @kumpera has offered a few explanations for why it's needed and XLA has a similar 'channel_id' field that iiuc is used for a similar purpose.

The gist is that you may want to allow more than one ProcessGroup (which owns real things like cuda streams) to exist for the same set of ranks, and let user code or compiler code be explicit about which one it's working with.  @kumpera can say more.

**Comment by kumpera** [comment]
> Well, idk what Will has in mind, but I would instead have repped it the other way: [1,2,3,4,5,6,7,8], stride=4 gives me [[1,2,3,4],[5,6,7,8]]

stride is probably unhelpful here, maybe `group_size` would be a better name.

**Comment by kumpera** [comment]
> Remind me what tag is?

Tag enables coordination of what PGs to use with the world outside of a compilation function, which is important for things like checkpointing or other eager affairs.

And within a function, it enables the same set of ranks to use different PGs and enable things like different stream priorities.

**Comment by ezyang** [comment]
so what, we have to parse a string? Would it be better to be an int?

**Comment by kumpera** [comment]
> so what, we have to parse a string? Would it be better to be an int?

The tag is just a key in a dictionary so either way works.
The advantage of a string is that it would be easer to read on graph dumps.

**Comment by 0x6b64** [comment]
Hi @wconstab, thank you so much creating an RFC for this work. A few questions for my clarity
1) Is it right to say that the ProcessGroup/Backend API will continue to stay the same (especially from a perspective of 3p vendor implementing custom collective backends)? Looks like there is a parallel development to refactor the ProcessGroup interface here: https://github.com/pytorch/pytorch/issues/86225

    The most noticeable part is that ProcessGroup will house a registry of backends each of which maps to the device for which they will work. At runtime, the actual collective backend if chosen depending on the device with which the input resides.

    Am I understanding this correctly that the `collective` API is a wrapping the call to ProcessGroup->Backend and holds onto the returned work item. I can see this is the implementation in `_all_reduce` in the WIP PR - https://github.com/pytorch/pytorch/pull/93379/files

    Also, there seems to be a concept of `ProcessGroupWrapper`'s in distributed_c10d. Is it right to say that doesn't need to be decomposed into chained operations?

2) I'm inferring that the DDP module will be updated to call the traceable_collectives allreduce instead of `dist.allreduce`, is this right?

3) Does the inclusion of the traceable collectives in the ddp graph immediately imply that DDP strategy used prior to DDPOptimizer will be resurrected && the emergent positioning of the collective calls will result in optimal compute/network overlap. Point 3 is phrased loosely, but I hope it communicates the general question I'm asking.

**Comment by wconstab** [comment]
For (1), we are thinking of building on top/alongside of existing processgroup.  So while we offer new python frontend apis (list of ranks, processgroup, dtensor.mesh), we can 'desugar' these representations to a canonical format based on list+int which we can trace into a graph.  Then backends (such as inductor) can translate them back into a real processgroup object and leverage the existing pg backends.

For (2, 3), we probably won't update the existing DDP codebase but we may offer some new way of getting a data-parallel model that is compiler friendly.  In general, beyond simple DDP style, we want to make the inductor backend more capable of optimizing comm+compute overlap and memory allocations, etc.  Thus it should become simpler to write a pretty naive 'ddp frontend' that then becomes optimal via the compiler backend.

**Comment by anj-s** [comment]
Does this API assume we need to know the ranks participating in the collectives?
```aten::collective(Tensor, *, str tag, int[] ranks, int stride) -> Tensor```
How can we get this information if we have not initialized the PG?

**Comment by wconstab** [comment]
My assumption is it is generally possible to code up your script in terms of one input WORLD_SIZE, which usually comes from ENV anyway, so you can get it without initializing distributed.  And definitely without initializing any pgs.

I do want to discuss a few related points further:
1) is it bad practice to just 'peek at ENV to get world size'?
2) should we tell people to init distributed before they get world_size?
2a) relatedly, should we init distributed (but not init any process groups) before tracing a distributed program?  Why?

3) how important is tracing a size-agnostic program?  Should we prioritize a way to trace a program that will work under any, say, power of 2 variant of WORLD_SIZE?  If so this seems hard with our canonical form of list<int> for ranks, even if we allowed list<symint> instead, since the size of that list is ultimately the thing we'd want to vary.

**Comment by anj-s** [comment]
> My assumption is it is generally possible to code up your script in terms of one input WORLD_SIZE, which usually comes from ENV anyway, so you can get it without initializing distributed. And definitely without initializing any pgs.

I think we can assume that. However I am just wondering 1) Are the ranks going to be passed to the SPMD API? 2) why not init distributed? 

> I do want to discuss a few related points further:
> 
> 1. is it bad practice to just 'peek at ENV to get world size'?
I like ENV vars however having a self contained script has its advantages. 

> 2. should we tell people to init distributed before they get world_size?
>    2a) relatedly, should we init distributed (but not init any process groups) before tracing a distributed program?  Why?
Yes, same question here. For initializing process groups we need to know the mesh dimensions over which we will be sharding our model. This might not happen until after the autoparallel profiling. However initializing a global PG seems fine to me. 

> 3. how important is tracing a size-agnostic program?  Should we prioritize a way to trace a program that will work under any, say, power of 2 variant of WORLD_SIZE?  If so this seems hard with our canonical form of list for ranks, even if we allowed list instead, since the size of that list is ultimately the thing we'd want to vary.
I think we need to know the cluster size to shard the model optimally. Given our sharding decision maybe based on cluster size, we might want to retrace. If we assume we don't want to reshard, is the tracing cost so high that we need to skip it?

**Comment by wconstab** [comment]
If we don't anticipate it being useful to trace a model (export it) and then later decide what the cluster size is, then we can just ignore that line of thought.  However, at least for simple things like DDP/FSDP use cases this seems like a pretty nice thing to be able to do.

I don't think tracing cost is that high relative to other aspects of distributed compilation.  It's really just for export and reuse that I think that could be useful.

**Comment by awgu** [comment]
> The semantics of these ops are that they return a real tensor, but you aren't allowed to access its data or storage.
...
> - It is valid to peek at the size() or stride() (or probably other metadata) of a tensor returned from a collective, but not its data.

What is the difference between the tensor "data" vs. "storage"?

**Comment by kumpera** [comment]
> > The semantics of these ops are that they return a real tensor, but you aren't allowed to access its data or storage.
> > ...
> > 
> > * It is valid to peek at the size() or stride() (or probably other metadata) of a tensor returned from a collective, but not its data.
> 
> What is the difference between the tensor "data" vs. "storage"?

One is the address of the tensor, the other are the values of each tensor element.

**Comment by kumpera** [comment]
> If we don't anticipate it being useful to trace a model (export it) and then later decide what the cluster size is, then we can just ignore that line of thought. However, at least for simple things like DDP/FSDP use cases this seems like a pretty nice thing to be able to do.
> 
> I don't think tracing cost is that high relative to other aspects of distributed compilation. It's really just for export and reuse that I think that could be useful.

I think we should decide whether we want to entertain all those would be nice scenarios that are not handled now or leave them for later and unblock further exploration on whether the current API even works for the narrow scenarios we already have.

While we believe the current design works for the current scenario of having distributed inited and ranks baked into the ops, we have no e2e validation of it.

**Comment by amithrm** [comment]
@wconstab Had a quick question about how we would impose control dependencies across the CC ops. Don't know if its relevant or makes sense to talk about it in the context of torch FX graph where the ordering is already established. I am asking this context from a PT/XLA view point where we insert tokens across CC ops. Will this be done at the backend level?

**Comment by wconstab** [comment]
> @wconstab Had a quick question about how we would impose control dependencies across the CC ops. Don't know if its relevant or makes sense to talk about it in the context of torch FX graph where the ordering is already established. I am asking this context from a PT/XLA view point where we insert tokens across CC ops. Will this be done at the backend level?

I'm don't think we need control deps at the fx layer and didn't design for it.  How are they inserted by torch-xla today?  if torch-xla can already insert them when tracing eager code, i suppose it can just treat the fx graph like eager code and insert tokens to 'freeze' the fx ordering.  Would that work?

**Comment by amithrm** [comment]
Yes, I think this should be possible. Its currently done under torch-xla using a separate token that is added into the computation. We can take the same approach

**Comment by alanwaketan** [comment]
Thanks, Will. Generally speaking this proposal should work for XLA at least for CPU/GPU/TPU. @amithrm Please chime-in for the Trainium side.

A few comments:
1. XLA normally doesn't need the PG for initialization as our runtime PJRT will figure the devices' topology for us. It's more like a bottom up approach instead of the torch.distributed's top down approach where the users tell the topology. @will-cromar to chime-in if I'm wrong. The current xla_backend for torch.distributed is just barely a layer to trace the cc ops into HLO. We do provide a https://github.com/pytorch/xla/blob/master/torch_xla/experimental/pjrt_backend.py#L14 for init. However it's just barely for following the torch.distributed's semantics, and it doesn't do anything real. **With the above speaking, I will imagine in this new world, XLA doesn't need to care about PG anymore.**

2. The ranks/stride things should work fine for us as we do have the concept of "[replica_groups](https://www.tensorflow.org/xla/operation_semantics#allreduce)" as well. However, not sure what's the tag mean for us?

3. Currently, PyTorch/XLA's cc ops support more parameters than the proposed aten::collective().
3.1. [channel_id](https://www.tensorflow.org/xla/operation_semantics#allreduce) from XLA. We currently don't lower it. Just FYI.
3.2. pin_layout, our own option, which is essential for us at this moment to force XLA to error out if we generate difference graphs on different processes. Detail lowering from LazyIR to HLO can be found here: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/cross_replica_reduces.cpp#L87
3.3 So my question is: **is it possible to expose some backend only options to cc ops**? Potentially it could be a void* or a dict, which serves as an extension point. For now, I guess we can always force pin_layout but Trainium might not like it.

cc @JackCaoG

**Comment by bdhirsh** [comment]
@wanchaol @wconstab do all of the functional collectives that we want exist at this point? (maybe we can go ahead and close this :) )

**Comment by mollon650** [comment]
I have the question about "do all of the functional collectives that we want exist at this point?"

**Comment by yifuwang** [comment]
Sharing some updates on functional collectives and seeking feedback.

We've been working on porting functional collectives to native ops and revamping their Inductor IRs. Most of these are improvements transparent to users, however there is an **API change** that we would like to propose as part of the revamp (discussed below).

### Transparent Improvements
Native collective ops
- Support collectives in AOTInductor
- Allow exported program containing collectives to run in native graph executors

Revamped Inductor IRs
- Make the Inductor IRs target agnostic
- Rely only on node dependencies for soundness
- Eliminate some unneccessary copies

### API Change
We are proposing to **remove the combination of ranks and tag as a supported group identifier in the collective APIs, and replace it with group_name**. This change will affect both the Python APIs and the aten IRs. Here are the rationales for this change:
- Capturing ranks as collective arguments pollutes and severely harms the readability of the captured graph.
- `(ranks, tag)` are unwieldy for communicator resolution. In certain cases, it even leads to exposed CPU overhead.
- It's confusing to users the purpose and the benefit of using this form of group identifier.
- The same information can be conveyed in a more concise manner (see @kumpera's comment: https://github.com/pytorch/pytorch/pull/109738#discussion_r1334538196)

### Migration Plan for Early Adoptors
The API change will affect both Python API users and backend implementors.

For Python API users, there will be a period during which `(ranks, tag)` is still supported group identifier, but using it leads to a warning. After this period, the users will need to adopt `group_name` as the group identifier.

For backend implementors, `group_name` will replace `(tag, ranks, group_size)` in the aten IR. The lowering logic will need to adapt to this change. Group membership information can be obtained via a dedicated API that takes `group_name`.

Specifically, the aten dispatcher signature (also the graph IR) will change from
```
all_reduce(Tensor self, str reduce_op, str tag, int[] ranks, int group_size) -> Tensor
```
to
```
all_reduce(Tensor self, str reduce_op, string group_name) -> Tensor
```

@alanwaketan please let me know if you have any concerns from XLA's perspective.

cc @wconstab @wanchaol @kumpera @voznesenskym @awgu @ezyang

**Comment by ezyang** [comment]
Ugggh does it have to be a string?

**Comment by yifuwang** [comment]
@ezyang it doesn't **need** to be a string. However since it's going to be an identifier anyway, I'm not sure I see the benefit of say an integer identifier. Could you elaborate the aversion to string?

**Comment by ezyang** [comment]
They have to be heap allocated, and comparisons on them have to be implemented with a loop

**Comment by alanwaketan** [comment]
> Sharing some updates on functional collectives and seeking feedback.
> 
> We've been working on porting functional collectives to native ops and revamping their Inductor IRs. Most of these are improvements transparent to users, however there is an **API change** that we would like to propose as part of the revamp (discussed below).
> 
> ### Transparent Improvements
> Native collective ops
> 
> * Support collectives in AOTInductor
> * Allow exported program containing collectives to run in native graph executors
> 
> Revamped Inductor IRs
> 
> * Make the Inductor IRs target agnostic
> * Rely only on node dependencies for soundness
> * Eliminate some unneccessary copies
> 
> ### API Change
> We are proposing to **remove the combination of ranks and tag as a supported group identifier in the collective APIs, and replace it with group_name**. This change will affect both the Python APIs and the aten IRs. Here are the rationales for this change:
> 
> * Capturing ranks as collective arguments pollutes and severely harms the readability of the captured graph.
> * `(ranks, tag)` are unwieldy for communicator resolution. In certain cases, it even leads to exposed CPU overhead.
> * It's confusing to users the purpose and the benefit of using this form of group identifier.
> * The same information can be conveyed in a more concise manner (see @kumpera's comment: [Native c10d_functional ops #109738 (comment)](https://github.com/pytorch/pytorch/pull/109738#discussion_r1334538196))
> 
> ### Migration Plan for Early Adoptors
> The API change will affect both Python API users and backend implementors.
> 
> For Python API users, there will be a period during which `(ranks, tag)` is still supported group identifier, but using it leads to a warning. After this period, the users will need to adopt `group_name` as the group identifier.
> 
> For backend implementors, `group_name` will replace `(tag, ranks, group_size)` in the aten IR. The lowering logic will need to adapt to this change. Group membership information can be obtained via a dedicated API that takes `group_name`.
> 
> Specifically, the aten dispatcher signature (also the graph IR) will change from
> 
> ```
> all_reduce(Tensor self, str reduce_op, str tag, int[] ranks, int group_size) -> Tensor
> ```
> 
> to
> 
> ```
> all_reduce(Tensor self, str reduce_op, string group_name) -> Tensor
> ```
> 
> @alanwaketan please let me know if you have any concerns from XLA's perspective.
> 
> cc @wconstab @wanchaol @kumpera @voznesenskym @awgu @ezyang

So far, we haven't implemented the ranks + group_size thing. Nothing really concerns me with this new approach. cc @YangFei1990 from AWS.

**Comment by seanlatias** [comment]
Hi @yifuwang, I'm also from AWS. After digging into the code, following are two issues and I would like to take some suggestions.

1. For supporting groups, how do Dynamo traces ProcessGroup itself. For example, when we use the API `torch.distributed.get_world_size(group=GP)`, `GP` currently cannot be traced successfully. In the new version, would that API also be changed with passing in a string name instead? I see there's another `_get_group_size_by_name()` API. This should be the right one to use?
2. With the new API (i.e., use string to refer to the group), we need to recover the actual ProcessGroup inside the C++ implementation (https://github.com/pytorch/xla/blob/master/torch_xla/csrc/cross_replica_reduces.cpp#L117). I guess this means we need to store a mapping between string name and the actual groups?
3. With the new functional ops, how will this API `torch.distributed.new_group()` be affected? Like what should be the right way to initialize a group? Can I use the `new_group()` API along with the `torch._C._distributed_c10d._register_process_group()` API?

cc @alanwaketan

**Comment by seanlatias** [comment]
Providing some error traces I got when running Dynamo.

```
Traceback (most recent call last):
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 490, in proxy_args_kwargs
    proxy_kwargs = {key: arg.as_proxy() for key, arg in kwargs.items()}
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 490, in <dictcomp>
    proxy_kwargs = {key: arg.as_proxy() for key, arg in kwargs.items()}
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/variables/base.py", line 282, in as_proxy
    raise NotImplementedError(str(self))
NotImplementedError: UserDefinedObjectVariable(ProcessGroupXla)

...
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 1162, in CALL_FUNCTION_KW
    self.call_function(fn, args, kwargs)
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 557, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/variables/torch.py", line 734, in call_function
    *proxy_args_kwargs(args, kwargs),
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 496, in proxy_args_kwargs
    raise unimplemented(
  File "/home/ubuntu/aws_neuron_venv_pytorch/lib/python3.8/site-packages/torch/_dynamo/exc.py", line 172, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_function args:  UserDefinedObjectVariable(ProcessGroupXla)
```

**Comment by wconstab** [comment]
RE (1)

> For supporting groups, how do Dynamo traces ProcessGroup itself. For example, when we use the API torch.distributed.get_world_size(group=GP), GP currently cannot be traced successfully. In the new version, would that API also be changed with passing in a string name instead? I see there's another _get_group_size_by_name() API. This should be the right one to use?

For this I think we can teach dynamo to evaluate `torch.distributed.get_world_size(group=GP)` during tracing and burn the resulting constant into the graph.  We already guard on the processgroup objects at runtime, iirc.  Pls open an issue for this one.

For (2, 3) IIUC this situation might be improved by @yifuwang's change, because now PG name is passed along with all the ops.  I didn't check the implementation though- presumably there is a mapping of name to obj already, which you could use too.

**Comment by seanlatias** [comment]
@wconstab thank you so much for the pointer. I do find that there's a `ProcessGroupVariable`. Now the issue is we are using `istype(value, (ProcessGroup, FakeProcessGroup))` to check. However, we in AWS are using XlaProcessGroup, which inherits from ProcessGroup. Do you think it makes sense to change the logic from `istype` to `isinstance`?

**Comment by seanlatias** [comment]
But it still errors out if I set `fullgraph=True` to `torch.compile()`. Maybe not setting the `fullgraph` option is the right solution because we don't want that part to be traced by Dynamo.

**Comment by yifuwang** [comment]
> Now the issue is we are using istype(value, (ProcessGroup, FakeProcessGroup)) to check. However, we in AWS are using XlaProcessGroup, which inherits from ProcessGroup. Do you think it makes sense to change the logic from istype to isinstance?

@seanlatias I think it should be fine. We can help with this in a separate issue.

> With the new API (i.e., use string to refer to the group), we need to recover the actual ProcessGroup inside the C++ implementation (https://github.com/pytorch/xla/blob/master/torch_xla/csrc/cross_replica_reduces.cpp#L117). I guess this means we need to store a mapping between string name and the actual groups?

There will be APIs for querying group info (e.g. membership) given group name. You will have to gather the required info at some point and make them available to the builders (e.g. via the bridge). I understand this will be more work, but it offers you flexibility to plumb through any info you need in any form (it would be impractical for us to capture rich info in op signature that satisfies all potential comm backends integrating with PT2).

Let us know if you think this will work, and if there's anything we can help to make it eaiser.

> With the new functional ops, how will this API torch.distributed.new_group() be affected? Like what should be the right way to initialize a group? Can I use the new_group() API along with the torch._C._distributed_c10d._register_process_group() API?

No. Process group initialization semantic will not be affected.

**Comment by seanlatias** [comment]
@yifuwang thanks for the clarification.
> @seanlatias I think it should be fine. We can help with this in a separate issue.

I'll create an issue for this. I also have some related questions and I'll post them in the same issue.
> Let us know if you think this will work, and if there's anything we can help to make it eaiser.

Do you have an example for this?

**Comment by yifuwang** [comment]
@seanlatias

> There will be APIs for querying group info (e.g. membership) given group name.
> Do you have an example for this?

It would roughly work like this:

```python
import os

import torch
import torch._C._distributed_c10d as c10d
import torch.distributed as dist
import torch.distributed._functional_collectives as funcol


def backend(gm, _):
    for node in gm.graph.nodes:
        if node.target == torch.ops._c10d_functional.all_reduce:
            group_name = node.args[2]
            # Get group ranks given group name
            pg = c10d._resolve_process_group(group_name)
            print(dist.get_process_group_ranks(pg))
    return gm


def func(t: torch.Tensor, group: dist.ProcessGroup) -> torch.Tensor:
    return funcol.all_reduce(t, "avg", group) + 1


if __name__ == "__main__":
    rank = int(os.environ["RANK"])
    local_rank = int(os.environ["LOCAL_RANK"])

    torch.cuda.set_device(f"cuda:{local_rank}")
    dist.init_process_group(backend="nccl")

    args = (torch.rand(20, 20, device="cuda"), dist.group.WORLD)
    torch.compile(func, fullgraph=True, backend=backend)(*args)
```

**Comment by tombousso** [comment]
> The collective ops are functional, but compilers may be able to convert them to inplace. They are asynchronous.

This creates a potential issue of large functional collective ops exhausting GPU memory, since they are not inplace.

1. Is there any support currently in inductor for optimizing the functional collectives to inplace?
2. If choosing to run a functional collective op in eager mode, is there any way to optimize to inplace?

**Comment by wconstab** [comment]
> @wconstab 
> 
> 
> 
> > The collective ops are functional, but compilers may be able to convert them to inplace. They are asynchronous.
> 
> 
> 
> This creates a potential issue of large functional collective ops exhausting GPU memory, since they are not inplace.
> 
> 
> 
> 1. Is there any support currently in inductor for optimizing the functional collectives to inplace?
> 
> 2. If choosing to run a functional collective op in eager mode, is there any way to optimize to inplace?

For (1), yes, inductor should reinplace if possible and generally optimize memory usage within the compiled region. 

For (2) not literally, if you call an FC in eager you get an FC along with its memory usage. (It would be a hit to peak memory but could be recouped if the tensor is discarded after some other operation.  But as a general pattern, we started trying to make torch dynamo able to trace native functional collective ops and reinterpret them as functional ones for tracing purposes- so if you want to keep using in place collectives in eager for performance reasons you may still have a path to compilation. 

@yifuwang or @yf225 may have more up to date information.

**Comment by ConnollyLeon** [comment]
Hi@wconstab, When will you support Async Allgather? It's commonly used in Megatron-LM sequence parallelism.

**Comment by YangFei1990** [comment]
> Hi@wconstab, When will you support Async Allgather? It's commonly used in Megatron-LM sequence parallelism.

Hi @ConnollyLeon could you shared more details for this Async allgather in Megatron-LM SP? Any pointer to the code/doc would be great. Thanks.

**Comment by ConnollyLeon** [comment]
> > Hi@wconstab, When will you support Async Allgather? It's commonly used in Megatron-LM sequence parallelism.
> 
> Hi @ConnollyLeon could you shared more details for this Async allgather in Megatron-LM SP? Any pointer to the code/doc would be great. Thanks.

Sure @YangFei1990 Sure, please check [here](https://github.com/NVIDIA/Megatron-LM/blob/80c7c6e936e6868fe251eabb079ae5d84ee28311/megatron/core/tensor_parallel/layers.py#L451) for all_gather and [here](https://github.com/NVIDIA/Megatron-LM/blob/80c7c6e936e6868fe251eabb079ae5d84ee28311/megatron/core/tensor_parallel/layers.py#L485) for reduce_scatter.

**Comment by xmfan** [comment]
can close this issue? funcols and dynamo rewrite have been in main for a while now. and there's a separate issue for documenting the API https://github.com/pytorch/pytorch/issues/113669
